Fix #16028 and #12647: Cache pip dependencies and make E2E tests use shared webpack build

U8NWXD opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #16028 and fixes #12647.
2. This PR does the following:
   * Enables caching of pip dependencies to reduce our susceptibility to network problems
   * Runs each E2E test as a separate job. This makes it easier to find E2E test runs of interest because one no longer has to keep track of which suite is in which workflow's job.
   * Uses a strategy matrix to auto-generate the job for each E2E test. This reduces duplicate code across our E2E test workflow files.
   * Before all the E2E jobs are started, download dependencies and compile webpack. Then make these prepared files available for each E2E test run, reducing overall runtime by avoiding redundant downloads and webpack compilation. Avoiding redundant downloads also reduces our susceptibility to network problems.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

Work remaining:

* [x] Confirm that we can pass webpack files between jobs as an artifact. Since building webpack is the bulk of our installation (~20 mins), reusing these is the most important step.
* [x] Confirm that dependencies in oppia_tools are correctly being persisted to the second job. Installing dependencies is the next most time-consuming step (~4 mins), and since we cache pip packages (new in this PR) and node modules, most of the savings will likely come from oppia_tools. To check this, make sure that the installing dependencies step in the second job is fast.
* [x] Use a [matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs) to auto-generate jobs for each e2e suite from a single `job` section in the workflow file. This will cut down on code duplication and let us run every e2e suite as its own job.

## Proof that changes are correct

### Proof that pip caching works

In the CI runs on this PR ([example](https://github.com/oppia/oppia/actions/runs/3520220855/jobs/5901504286#step:6:3332)) we load pip-tools from the cache:

```
Collecting pip-tools==6.6.2
  Using cached pip_tools-6.6.2-py3-none-any.whl (47 kB)
```

In contrast, in a CI run from develop ([example](https://github.com/oppia/oppia/actions/runs/3521701243/jobs/5903838435#step:6:3382)), we download pip-tools:

```
Collecting pip-tools==6.6.2
  Downloading pip_tools-6.6.2-py3-none-any.whl (47 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.7/47.7 kB 2.0 MB/s eta 0:00:00
```

### Proof that third-party dependencies transfer

Notice that in [this CI run](https://github.com/U8NWXD/oppia/actions/runs/3545047863/jobs/5953186969) the install-oppia-dependencies step takes only ~2 mins instead of the usual ~4 mins. This speed-up is because third-party dependencies like the google cloud SDK don't have to be downloaded--they were loaded from the zip file.

### Proof that upload/download is worthwhile

Comparing [this new CI run](https://github.com/U8NWXD/oppia/actions/runs/3545047863/jobs/5953186969) to an [old run](https://github.com/oppia/oppia/actions/runs/3536962376/jobs/5936508229) we can see the impact of these changes on overall runtime. Note that we previously had 17 E2E test workflows, and we have 43 E2E test suites.

| Step | Duration | Times needed before | Total Duration Before | Times needed now | Total Duration Now |
|---|---|---|---|---|---|
| Creating zip file of build files | 3:39 | 0 | 0:00 | 1 | 3:39 |
| Uploading zip file | 7:12 | 0 | 0:00 | 1 | 7:12 |
| Downloading zip file | 0:39 | 0 | 0:00 | 43 | 27:57 |
| Unzipping build files | 0:46 | 0 | 0:00 | 43 | 32:58 |
| Compiling webpack | 25:36 | 17 | 435:12 | 1 | 25:36 |
| install-oppia-dependencies (no zip file) | 5:24 | 17 | 91:48 | 1 | 5:24 |
| install-oppia-dependencies (with zip file) | 1:56 | 0 | 0:00 | 43 | 83:08 |
| **Total**| | | **527:00** | | **185:54** |

Therefore even though we spin up more jobs with this approach and transfering build files as an artifact takes time, the benefits from only having to download dependencies and build webpack once mean that this transition will save us time overall.

### Proof that all tests work

See CI runs: https://github.com/U8NWXD/oppia/actions/runs/3545519837

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

U8NWXD committed:
Enable pip caching on lint.yml as a test

U8NWXD committed:
Pass build files between workflow jobs

U8NWXD committed:
Zip build files before uploading

U8NWXD committed:
For testing, delete e2e workflows not being developed currently

U8NWXD committed:
Fix artifacts upload path

U8NWXD committed:
Add commands to debug build_files.zip location

U8NWXD committed:
Add ls -la debugging

U8NWXD committed:
Add cd that I accidentally removed

U8NWXD committed:
Use working-directory instead of cd

U8NWXD committed:
Specify shell

U8NWXD committed:
Update download path to be parent folder

U8NWXD committed:
Merge branch 'develop' into cache-pip

U8NWXD committed:
Add debugging ls -la statements

U8NWXD committed:
Fix zip to add files recursively

U8NWXD committed:
Add webpack bundles to artifact

U8NWXD committed:
Include more build files in artifact, incl. app.yaml

U8NWXD committed:
Remove comments and line breaks from zip command

U8NWXD committed:
Preserve symlinks when zipping

U8NWXD committed:
Include backend_prod_files/ in zip

U8NWXD committed:
Implement strategy matrix for e2e tests

U8NWXD committed:
Rename e2e workflow yaml now that we only have one

U8NWXD committed:
Remove colon from name

U8NWXD committed:
Revert "For testing, delete e2e workflows not being developed currently"

This reverts commit 11bf75ad16fefb0645f717bd55712977257917d6.

U8NWXD committed:
Delete old e2e workflows

U8NWXD committed:
Fix e2e captured in CI check

U8NWXD committed:
Fix check_e2e_tests_are_captured_in_ci tests

seanlip commented:
Thanks @U8NWXD! The code changes look great to me.

Could you please give me the exact instructions for what to change in our github settings before I merge this into develop? Then I'll LGTM this and merge it.

Thanks a lot for the detailed analysis, too! I had started writing out a question about time savings and then realized you had already answered it comprehensively in the description :)

U8NWXD commented:
@seanlip sure! You might not be able to mark these new CI checks as required until they've run on develop. We also probably don't want to remove the old e2e workflows from the required list until we've added the new ones. Therefore, I think this is the best order of operations:

1. Force-merge this PR after confirming that the only non-passing CI checks are those that we'll be removing in step 2 below.

2. Wait for CI checks from this PR to start running so that they appear in the list of available checks to make required.

3. Add the CI checks of the form `End-to-End tests / e2e_test (ubuntu-22.04, SUITE_NAME)` to the list of required checks. There are 43 of these checks, one for each e2e suite, and they're listed in the CI checks that ran for this PR. Construct these checks by replacing SUITE_NAME with each of the suite names below:

          - collections
          - accessibility
          - additionalEditorFeatures
          - additionalEditorFeaturesModals
          - additionalPlayerFeatures
          - adminPage
          - blogDashboard
          - blog
          - checkpointFeatures
          - classroomPage
          - classroomPageFileUploadFeatures
          - collections
          - contributorDashboard
          - coreEditorAndPlayerFeatures
          - creatorDashboard
          - diagnosticTestPage
          - embedding
          - explorationFeedbackTab
          - explorationImprovementsTab
          - explorationHistoryTab
          - explorationStatisticsTab
          - explorationTranslationTab
          - extensions
          - featureGating
          - fileUploadExtensions
          - fileUploadFeatures
          - learner
          - learnerDashboard
          - library
          - navigation
          - playVoiceovers
          - preferences
          - profileFeatures
          - profileMenu
          - publication
          - skillEditor
          - subscriptions
          - topicsAndSkillsDashboard
          - topicAndStoryEditor
          - topicAndStoryEditorFileUploadFeatures
          - topicAndStoryViewer
          - users
          - wipeout

4. Remove all the `e2e_* (ubuntu-22.04)` CI checks from the list of required checks. These workflows are removed by this PR.

I've added the `require post-merge sync to HEAD` label so everyone will be notified that they need to merge in develop to their PRs to get the new E2E tests.

seanlip commented:
Thanks @U8NWXD, will do! Just an FYI though, we are getting a few last-minute GSoC PRs merged, so I'll take a look at this in 24 hours or so, and not do anything for now in order to avoid churn on develop.

seanlip commented:
OK, doing step 1 now!

seanlip commented:
@U8NWXD Steps 3 and 4 of [this comment](https://github.com/oppia/oppia/pull/16629#issuecomment-1328060351) are now done! However, I have two questions:

(a) In step 3, "collections" is included twice. Does there mean that there should be 42 workflows added, or that there should still be 43 because one of them is mistyped? (I've only added 42 so far.)

(b) In step 4, am I correct that `e2e_coverage_and_typescript_tests (ubuntu-22.04)` should be left as-is and not made un-required, or whether I should drop it as well? Currently, I have gone with the first one.

Thanks!

U8NWXD commented:
a) Oops, collections should only appear once. I'll open a PR to remove the duplicate entry.

b) Yes, please leave `e2e_coverage_and_typescript_tests` un-changed. That job has not changed.

U8NWXD commented:
Opened https://github.com/oppia/oppia/pull/16645 to remove the duplicate `collections` entry
